pkg/util/deployment/deployment.go

```diff
@@ -236,6 +236,10 @@ func addHashKeyToRSAndPods(deployment *extensions.Deployment, c clientset.Interf
 		return updatedRS, nil
 	}
 
+	if err = waitForPodsHashPopulated(c, namespace, updatedRS.Name); err != nil {
```

Comment:
Is it possible that the rs's status.fullyLabeledReplicasCount is stale before this check? For example, before we update the rs's pod template, that value already matches spec.replicas. If so, wait for it to become `0` before step 2. If not, note this in the comment.